### PR TITLE
Refactor Eks Create Cluster Operator code

### PR DESCRIPTION
**Pull request by syedahsn**
The `EksCreateClusterOperator` is capable of creating a Fargate Profile or Nodegroup that is associated with the newly created cluster. The code for creating the Fargate Profile and Nodegroup is the same as the code in `EksCreateFargateProfileOperator` and `EksCreateNodegroupOperator` respectively. 
This PR refactors the code to reuse some of the code to decrease code duplication. This step also simplifies adding deferrable mode to these operators.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/providers/amazon/aws/operators/eks.py [hunk 1]
```diff
@@ -59,7 +59,54 @@
 FARGATE_FULL_NAME = "AWS Fargate profiles"
 
 
-class EksCreateClusterOperator(BaseOperator):
```
Comment (phanikumv) [inline airflow/providers/amazon/aws/operators/eks.py:62]: Dont we need deprecation warning here, and call the new operator if we remove this class? Otherwise existing DAGs using the `EksCreateClusterOperator` will break right?

Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/eks.py:62]: I'm not removing the class, just making it inherit from a different class. It shouldn't affect any existing DAGs.

File: airflow/providers/amazon/aws/operators/eks.py [hunk 2]
```diff
@@ -184,12 +235,12 @@ def execute(self, context: Context):
                     )
                 )
 
-        eks_hook = EksHook(
+        self.eks_hook = EksHook(
             aws_conn_id=self.aws_conn_id,
             region_name=self.region,
         )
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/eks.py]: Now that you have a cool base class, you can add an __init__ to it and create the hook inside of that so that all the child classes don't need to create it each time.

File: airflow/providers/amazon/aws/operators/eks.py [hunk 3]
```diff
@@ -393,51 +416,41 @@ def __init__(
     ) -> None:
         self.cluster_name = cluster_name
         self.pod_execution_role_arn = pod_execution_role_arn
+        self.fargate_pod_execution_role_arn = pod_execution_role_arn
         self.selectors = selectors
+        self.fargate_selectors = selectors
         self.fargate_profile_name = fargate_profile_name
         self.create_fargate_profile_kwargs = create_fargate_profile_kwargs or {}
-        self.wait_for_completion = wait_for_completion
+        self.wait_for_completion = False if deferrable else wait_for_completion
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/eks.py:473]: Why is this only done on this operator but not the previous two?

File: airflow/providers/amazon/aws/triggers/eks.py [hunk 4]
```diff
@@ -33,49 +33,52 @@ class EksCreateFargateProfileTrigger(BaseTrigger):
 
     :param cluster_name: The name of the EKS cluster
     :param fargate_profile_name: The name of the fargate profile
-    :param poll_interval: The amount of time in seconds to wait between attempts.
-    :param max_attempts: The maximum number of attempts to be made.
+    :param waiter_delay: The amount of time in seconds to wait between attempts.
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/triggers/eks.py:36]: Why all the renaming? The PR description doesn't mention this and it seems unrelated to factoring out the create compute method?

File: airflow/providers/amazon/aws/operators/eks.py [hunk 5]
```diff
@@ -59,7 +59,52 @@
...
 
-class EksCreateClusterOperator(BaseOperator):
+class BaseEksCreateOperator(BaseOperator):
+    """
+    Base class for EKS Operators that create cluster resources. This includes
+    EksCreateClusterOperator, EksCreateNodeGroupOperator and EksCreateFargateProfileOperator.
+
+    """
+
+    def create_compute(cls):
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/eks.py]: This is an instance method, can you s/cls/self/g in all places for this method?

File: airflow/providers/amazon/aws/operators/eks.py (hunk 2 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/eks.py]: Nice, that's a good idea. Thanks!

File: airflow/providers/amazon/aws/operators/eks.py (hunk 3 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/eks.py:473]: Because as of right now, this is the only Operator that has the deferrable property. Once deferrable mode is implemented for the other 2 operators, I'll update `wait_for_completion` to match this one.

File: airflow/providers/amazon/aws/triggers/eks.py (hunk 4 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/triggers/eks.py:36]: I realized I've been naming these parameters incorrectly for the past few Triggers, and I was hoping to fix them. Considering the `EksCreateFargateProfileTrigger` got merged very recently, I was hoping to do this without doing a  deprecation warning. If it is problematic to include it as part of this PR, I can take it out.

File: airflow/providers/amazon/aws/operators/eks.py (hunk 3 above)
Comment (o-nikolas) [reply airflow/providers/amazon/aws/operators/eks.py:473]: Ahh, gotcha, makes sense :+1:

File: airflow/providers/amazon/aws/triggers/eks.py (hunk 4 above)
Comment (o-nikolas) [reply airflow/providers/amazon/aws/triggers/eks.py:36]: > got merged very recently, I was hoping to do this without doing a deprecation warning

This should be fine since we haven't done a provider release yet

 > If it is problematic to include it as part of this PR, I can take it out.
 
In a perfect world I think a separate PR would have been best, but it's not a huge deal, so let's keep it here for this time :+1:

File: airflow/providers/amazon/aws/operators/eks.py (hunk 1 above)
Comment (vincbeck) [reply airflow/providers/amazon/aws/operators/eks.py:62]: Yes Github makes it confusing but look at line `107`, `EksCreateClusterOperator` is still there

File: airflow/providers/amazon/aws/operators/eks.py (hunk 2 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/eks.py]: Made the changes

File: airflow/providers/amazon/aws/operators/eks.py [hunk 6]
```diff
@@ -184,12 +238,12 @@ def execute(self, context: Context):
                     )
                 )
 
-        eks_hook = EksHook(
+        self.eks_hook = EksHook(
             aws_conn_id=self.aws_conn_id,
             region_name=self.region,
         )
```
Comment (o-nikolas) [inline airflow/providers/amazon/aws/operators/eks.py]: This can be removed now right?

Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/eks.py]: sorry I missed that. Changing it now

File: airflow/providers/amazon/aws/operators/eks.py [hunk 7]
```diff
@@ -59,7 +59,55 @@
 FARGATE_FULL_NAME = "AWS Fargate profiles"
 
 
-class EksCreateClusterOperator(BaseOperator):
+class BaseEksCreateOperator(BaseOperator):
+    """
+    Base class for EKS Operators that create cluster resources. This includes
+    EksCreateClusterOperator, EksCreateNodeGroupOperator and EksCreateFargateProfileOperator.
+
+    """
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/operators/eks.py]:
```suggestion
    """Base class for EKS Operators that create cluster resources.

    This includes EksCreateClusterOperator, EksCreateNodeGroupOperator and
    EksCreateFargateProfileOperator.
    """
```

I’m not sure we should have the second line though. The list of classes doesn’t really add much useful information for maintenance, and is very easy to get stale.

File: airflow/providers/amazon/aws/operators/eks.py [hunk 8]
```diff
@@ -59,7 +59,55 @@
...
+
+    def __init__(self, **kwargs):
+        self.eks_hook = EksHook(aws_conn_id=self.aws_conn_id, region_name=self.region)
+        super().__init__(**kwargs)
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/operators/eks.py]: Any reason the hook is instantiated before the `super()` call? I believe the majority preference is to do it after the super call. (I don’t think the style matters here, but it may introduce unnecessary confusion.)

File: airflow/providers/amazon/aws/operators/eks.py [hunk 9]
```diff
@@ -59,7 +59,55 @@
...
+
+    def create_compute(self):
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/operators/eks.py]:
```suggestion
    def create_compute(self) -> None:
```

And now Mypy would start to prompt you adding variable definitions on the class (which is good)

File: airflow/providers/amazon/aws/operators/eks.py (hunk 7 above)
Comment (uranusjr) [reply airflow/providers/amazon/aws/operators/eks.py]: Also maybe a `:meta private:` would be a good idea? I don’t think we expect users to call this class directly. (Or rename this class to `_BaseEksCreateOperator`)

**Review by vincbeck — APPROVED** [review]

File: airflow/providers/amazon/aws/operators/eks.py (hunk 8 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/eks.py]: No particular reason. I didn't really pay attention to the order :sweat_smile:

File: airflow/providers/amazon/aws/operators/eks.py (hunk 9 above)
Comment (syedahsn) [reply airflow/providers/amazon/aws/operators/eks.py]: I made this change, and it ended up having a bigger impact than I anticipated. I had to add a bunch of parameters used in the 3 child classes. I think I've got everything ready. I need to do some testing with how templating works here. Right now, `_BaseEksCreateOperator` doesn't have any templated fields, but because the child classes allow templating for some parameters, I think I have to do the same for `_BaseEksCreateOperator`. 

@uranusjr what do you think?

**Review by uranusjr — APPROVED** [review]

**Comment by uranusjr** [comment]
Looks like there’s a coverage error. Maybe caused by the base class name change?

**Review by o-nikolas — APPROVED** [review]
